File: Data Structures & Algorithms/mergeSort/submission-0.js
/** Pair class to store key-value pairs */
// class Pair {
//   /**
//    * @param {number} key The key to be stored in the pair
//    * @param {string} value The value to be stored in the pair
//    */
//   constructor(key, value) {
//       this.key = key;
//       this.value = value;
//   }
// }
class Solution {
    /**
     * @param {Pair[]} pairs
     * @returns {Pair[]}
     */
    mergeSort(pairs) {
		if (pairs.length <= 1) {
			return pairs;
		}

		const middleIndex = Math.floor(pairs.length / 2);
		const firstHalf = this.mergeSort(pairs.slice(0, middleIndex));
		const secondHalf = this.mergeSort(pairs.slice(middleIndex));

		const result = [];
		let firstI = 0;
		let secondI = 0;
		while (firstI < firstHalf.length || secondI < secondHalf.length) {
			if (firstI >= firstHalf.length) {
				result.push(secondHalf[secondI]);
				secondI++;
				continue;
			}
			if (secondI >= secondHalf.length) {
				result.push(firstHalf[firstI]);
				firstI++;
				continue;
			}

			if (firstHalf[firstI].key <= secondHalf[secondI].key) {
				result.push(firstHalf[firstI]);
				firstI++;
				continue;
			} else {
				result.push(secondHalf[secondI]);
				secondI++;
				continue;
			}
		}

		return result;
	}
}

File: Data Structures & Algorithms/insertionSort/submission-0.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
	std::vector<std::vector<Pair>> insertionSort(std::vector<Pair>& pairs) {
		std::vector<Pair> result{};
		std::vector<std::vector<Pair>> resultHistory{};

		for (size_t i = 0; i < pairs.size(); i++) {
			for (size_t j = 0; j <= result.size(); j++) {
				if (j == result.size()) {
					result.push_back(pairs[i]);
					break;
				}

				if (result[j].key > pairs[i].key) {
					result.insert(result.begin() + j, pairs[i]);
					break;
				}
			}

			std::vector<Pair> log = result;
			for (size_t z = result.size(); z < pairs.size(); z++) {
				log.push_back(pairs[z]);
			}

			resultHistory.push_back(log);
		}

		return resultHistory; 
	}
};
